### ISSUE TITLE
Helper methods for the creation of masks

### ISSUE THREAD

**Issue by registerrier** [gammapy/gammapy#2869]
At the moment, we have several methods that can help create masks in gammapy. They belong to `Geom` or `WcsGeom`:
- `WcsGeom.region_mask()` which takes a list of region and returns a boolean array with the same dimension as the `Geom`.
- `Geom.energy_mask(min=, emax=)` which takes two energy bounds and returns a boolean array with the same dimension as the `Geom`. Since it is defined on `Geom` it works both with `WcsMap` and `RegionMap`.

This two methods probably cover most of the use cases we are dealing with. The more obvious other case to me is when a user wants to create a mask from another `Map` applying some thresholding (e.g. on a minimum exposure or maximal alpha). This you can do easily with `mymap.data > value` for instance.

The `Dataset` mask handling rather requires `Map` objects rather than arrays. This can be confusing (e.g. one cannot save a `MapDataset` that has an array defined for `mask_fit`). Hence the three options below:
0. Stay as is. 
1. Modify the  methods on `Geom` to return full `Map` objects to be more compliant with what is expected or `Dataset` mask? and then we might add a method on `Map` to make a threshold mask e.g. `Map.make_clip_mask(min_val=None, max_val=None, inside=True)` that would return a `Map` with the correct shape.
2. Add an API on the `Dataset` itself for the user to add his own `mask_fit`. Basically this would simply wrap over the `Geom` based methods. This should simplify the life of the user at the cost of another layer. After all, a user will have to interact a lot with `Datasets`, and should not be required to know the `Geom` API beyond `create`. 

Any opinion @adonath @AtreyeeS @Bultako @QRemy @bkhelifi ?

**Comment by adonath** [comment]
@registerrier We could also try to  implement the `___lt__` and `__gt__` and boolean operators for maps, such that the following just works:
```
mask = (my_map > 5) & (my_map < 10)
``` 
I agree we could directly return `Map` objects in the `Geom.energy_mask()` methods. On the other hand the difference is just one line of code...which I think is still acceptable.
```
data = geom.energy_mask()
maks = Map.from_geom(geom=geom, data=data) 
```

**Comment by AtreyeeS** [comment]
Thanks @registerrier for bringing this up! I would prefer something along the lines of option (2) that you suggest. I feel that users will have to interact with `mask_fits` a lot at the data fitting stage, and it is nice to have something at the `Datasets` level.

**Comment by adonath** [comment]
I'm definitely against adding a new API on the `Dataset` level for mask creation. The options for mask creation are endless and we just cannot support and maintain all the options to create masks. We should teach users Numpy instead or at least offer a Numpy compatible API for `Map` (maybe the proposal I put above is an acceptable compromise...).  The dataset API is already large enough and with the `.mask_fit` attribute we have a well defined interface. I also think hat users have to know the `Map` API. The goal should not be to make everything as simple as possible but to document it well instead!

I'm honestly against adding a `make_clip_mask(min_val=None, max_val=None, inside=True)` method either. Creating this mask is simple Numpy syntax and every Gammapy user should know how to do this. I would always prefer a docs example over a new and different (than Numpy) API.

**Comment by adonath** [comment]
Just to add to my comments above: I think `.energy_mask()` mask is somehow a special case, because we can implement a convention on how to compute the mask. If we leave it to the user in this case we will get all inconsistencies between using energy edges, centers, open and closed intervals.

**Comment by QRemy** [comment]
I agree with Axel, it's simple enough with Numpy, we don't have to create an extra layer of complexity in gammapy.

**Comment by adonath** [comment]
Thinking about it once more, it would be inconsistent to offer a `Map` factory method on the Geom object. I.e. make `.energy_mask()`, and `.region_mask()` return a `Map` object. The logic we implemented is the other way around, namely that a `Map` is created from a `Geom` object and not by a `Geom` object. I'm pretty sure this is the reason, why we made the functions return a `ndarray` in first place. 

I still think implementing the boolean operators for `Map` is a good option. It will be useful and simplify code in the other places as well...

**Comment by AtreyeeS** [comment]
Just to be clear, when you talk of `mask = (my_map > 5)`, you mean `np.where(my_map.data > 5)` ?

One of the reasons this discussion came about is because while doing some HESS analysis, I find that is is very necessary to be able to make integrated maps above certain thresholds. This can be doable by appropriately modifying the `mask_fit`, and it would be nice to have some convenience method to do so.

**Comment by QRemy** [comment]
This is something that can be explained in an how to like this:

Defining or modifying a mask_fit can be done in one line:
```python
geom = dataset.counts.geom
dataset.mask_fit=Map.from_geom(geom, data=mask)
dataset.mask_fit.data &= other_mask
```
Here are some mask examples (among endless possibilities) : 
Add a margin in (l,b) or select a sub-region for the fit
```python
coords = dataset.counts.geom.get_coord()
mask = ((coords["lon"]>=lmin) & (coords["lon"]<=lmax)
              & (coords["lat"]>=bmin) & (coords["lat"]<=bmax)
```
Limit energy range
```python
mask = (coords["energy"]>=Ethr))
```

Filter based on a map value, here we define a mask from the HGPS significance map and reproject it to our dataset coordinates
```python
significance = Map.read('hgps_map_significance_0.1deg_v1.fits')
mask_hgps = Map.from_geom(significance.geom, data = (significance.data > 3))
mask =  Map.from_geom(geom)
mask.data = ~mask.data.astype(bool) # change default  to True
coords2d =  mask.sum_over_axes().geom.get_coord()
mask2d = mask_hgps.get_by_coord(coords2d) # reproject
mask.data &= mask2d # this works even if mask is 3d
```

**Comment by adonath** [comment]
@AtreyeeS The `np.where()` call is never needed to create masks. A boolean mask can be just created using e.g. `mask = data > 5` and be used for indexing like `data[mask] = 15`.

**Comment by adonath** [comment]
@QRemy Here is a proposal on how we could extend our code to simplify your last example:
 
```python3
significance = Map.read('hgps_map_significance_0.1deg_v1.fits')
mask_hgps = Map.from_geom(significance.geom, data=(significance.data > 3))

mask =  Map.from_geom(data=1, geom, dtype=bool)
mask &= mask_hgps.reproject_to_geom(geom, interp="nearest")
```

This would require the following changes to Gammapy:
- Implement boolean operators for `Map`
- Implement a `reproject_to_geom()` convenience method
- Pass and use `dtype` in `Map.from_geom()`

**Comment by registerrier** [comment]
So far, we haven't implemented a `reproject_to_geom()` method. 
It is important to remember that it would be used frequently for other usages. So one has to be aware of the possible options regarding flux conservation.
Maybe a specific function such as `reproject_mask_to_geom` could be enough?

Anyway, we should implement boolean operators and comparators for `Map`:
`mask_hgps = significance>3` is even easier.
Do you want to do it @QRemy , @AtreyeeS ? Otherwise I can do it.

**Comment by registerrier** [comment]
The proposal above is implemented in PR #2885

**Comment by adonath** [comment]
Now we also have support for boolean operations for `Map` like `map_1 & map_2`. In addition we now have `Map.interp_to_geom()`as well, so I think there is nothing left to do here...
